network/betweenness.go

```diff
@@ -87,3 +93,57 @@ func Betweenness(g graph.Graph) map[int]float64 {
 
 	return cb
 }
+
+// BetweennessWeighted returns the betweenness centrality for nodes in the weighted
+// graph g used to construct the given shortest paths.
+//
+//  C_B(v) = \sum_{s ≠ v ≠ t ∈ V} (\sigma_{st}(v) / \sigma_{st})
+//
+func BetweennessWeighted(g graph.CostGraph, p search.ShortestPaths) map[int]float64 {
+	nodes := g.NodeList()
+	cb := make(map[int]float64, len(nodes))
+	for i, s := range nodes {
+		for j, t := range nodes {
+			if i == j {
+				continue
+			}
+			d := p.Weight(s, t)
+			if math.IsInf(d, 0) {
+				continue
+			}
+
+			sID := s.ID()
+			tID := t.ID()
+
+			// If we have a unique path, don't do the
+			// extra work needed to get all paths.
+			path, _, unique := p.Between(s, t)
+			if unique {
+				for _, v := range path {
+					if vID := v.ID(); vID == sID || vID == tID {
+						continue
+					}
+					// For undirected graphs we double count
+					// passage though nodes. This is consistent
+					// with Brandes' algorithm's behaviour.
+					cb[v.ID()]++
+				}
+				continue
+			}
+
+			// Otherwise iterate over all paths.
+			paths, _ := p.AllBetween(s, t)
+			stFrac := 1 / float64(len(paths))
```

Comment:
If there is no path between s and t, won't there be a division by 0? Or is the weight always infinite in such case at line 111?